README.md
# bbg-api

FIXME: description

## Installation

Download from http://example.com/FIXME.

## Usage

FIXME: explanation

    $ java -jar bbg-api-0.1.0-standalone.jar [args]

## Options

FIXME: listing of options this app accepts.

## Examples

...

### Bugs

...

### Any Other Sections
### That You Think
### Might be Useful

## License

Copyright © 2022 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2022-02-14
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2022-02-14
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://sourcehost.site/your-name/bbg-api/compare/0.1.1...HEAD
[0.1.1]: https://sourcehost.site/your-name/bbg-api/compare/0.1.0...0.1.1


test/bbg_api/core_test.clj
(ns bbg-api.core-test
  (:require [clojure.test :refer :all]
            [bbg-api.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/bbg_api/db.clj
(ns bbg-api.db
  (:require [clojure.xml :as xml]
            [clojure.pprint :as pp]))
;; 
;; Fields accessors from API XML
;; 
(defn- game-id [game]
  (get-in game [:attrs :objectid]))

(defn- game-my-rating [collection-game]
  (let [rating (-> collection-game
                   :content
                   (nth 4)
                   :content
                   first
                   :attrs
                   :value
                   read-string)]
    (if (number? rating) rating nil)))

(defn- game-rating [collection-game]
  (let [rating (-> collection-game
                   :content
                   (nth 4)
                   :content
                   first
                   :content
                   (nth 2)
                   :attrs
                   :value
                   read-string)]
    (if (number? rating) rating nil)))

(defn- game-name [collection-game]
  (-> collection-game
      :content
      first
      :content
      first))

(defn- game-attributes [collection-game]
  (-> collection-game
      :content
      (nth 4)
      :attrs))

(defn- game-attribute [collection-game]
  (fn [key]
    (let [attr (game-attributes collection-game)
          value (attr key)]
      (if value (read-string value) nil))))

;; (defn game-playingtime [collection-game]
;;   ((game-attribute collection-game) :playingtime))

;; (defn game-maxplayers [collection-game]
;;   ((game-attribute collection-game) :maxplayers))

;; (defn game-minplayers [collection-game]
;;   ((game-attribute collection-game) :minplayers))

;; 
;; Game api
;; 
(defn api-read-game [game-id]
  (Thread/sleep 1000)
  (-> (xml/parse (str "https://boardgamegeek.com/xmlapi/boardgame/" game-id))
      :content
      first))

(defn get-games-and-write-to-file [collection]
  (spit "resources/games.clj"
        (reduce
         #(assoc %1 (get-in %2 [:attrs :objectid]) %2)
         {}
         (map api-read-game (map game-id collection)))))

(defn read-games-from-file []
  (read-string (slurp "resources/games.clj")))

;; 
;; Collection API 
;; 
(defn fetch-collection-and-write-to-file [user-name]
  (spit "resources/collection.clj"
        (xml/parse (str "https://boardgamegeek.com/xmlapi/collection/" user-name))))

(defn read-collection-from-file []
  (:content (read-string (slurp "resources/collection.clj"))))

;; 
;; Functions for numbers of players
;; 
(defn- polls-with-num-of-players-for-game [game]
  (let [tag-list (game :content)
        tag-poll (filter (fn [x] (= (x :tag) :poll)) tag-list)
        recommended (filter (fn [x] (= (get-in x [:attrs :name]) "suggested_numplayers")) tag-poll)]
    ((first recommended) :content)))

(defn- votes-best-rating-per-players [data]
  (let [players (get-in data [:attrs :numplayers])
        total-votes (apply + (map (fn [x] (read-string (get-in x [:attrs :numvotes]))) (data :content)))
        best-votes (read-string (get-in (first (data :content)) [:attrs :numvotes]))
        best-perc (if (= 0 total-votes) 0 (double (/ best-votes total-votes)))

        recommended-votes (read-string (get-in (second (data :content)) [:attrs :numvotes]))
        recommended-perc (if (= 0 total-votes) 0 (double (/ recommended-votes total-votes)))
        not-recommended-votes (read-string (get-in (last (data :content)) [:attrs :numvotes]))
        not-recommended-perc (if (= 0 total-votes) 0 (double (/ not-recommended-votes total-votes)))]
    {:players players
     :best-votes best-votes :best-perc best-perc
     :recommended-votes recommended-votes :recommended-perc recommended-perc
     :not-recommended-votes not-recommended-votes :not-recommended-perc not-recommended-perc}))


;; 
;; API - DB
;; 
(defn- collection-game->game
  [games collection-game]
  (let [game-id (game-id collection-game)
        votes (map votes-best-rating-per-players
                   (polls-with-num-of-players-for-game (games game-id)))]
    {:id game-id
     :name (game-name collection-game)
     :rating (game-rating collection-game)
     :my-rating (game-my-rating collection-game)
     :minplayers ((game-attribute collection-game) :minplayers)
     :maxplayers ((game-attribute collection-game) :maxplayers)
     :playingtime ((game-attribute collection-game) :playingtime)
     :votes votes}))

(defn- make-db-from-collection-and-games
  [collection games]
  (let [all (map #(collection-game->game games %) collection)
        all-db (reduce
                #(assoc %1 (:id %2) %2)
                {} all)]
    (spit "resources/db.clj" (with-out-str (pp/pprint all-db)))))

(defn make-db
  []
  (make-db-from-collection-and-games
   (read-collection-from-file)
   (read-games-from-file)))

(defn read-db
  []
  (read-string (slurp "resources/db.clj")))

(comment
  ;; (def collection (g/read-collection-from-file))
  ;; (def games (g/read-games-from-file))
  ;; (make-db-from-collection-and-games collection games)

  (make-db)

  (def db (read-db))
  (db "25613")
  (pp/pp)
  (def collection (vals db))

  (take 2 collection)
  (clojure.pprint/pp)
  ;
  )

src/bbg_api/writefiles.clj
(ns bbg-api.writefiles
  (:require [bbg-api.core :as g]
            [clojure.pprint :as pp]))

(defn collection-game->game
  [games collection-game]
  (let [game-id (g/game-id collection-game)
        votes (map g/votes-best-rating-per-players
                   (g/polls-with-num-of-players-for-game (games game-id)))]
    {:id game-id
     :name (g/game-name collection-game)
     :rating (g/game-rating collection-game)
     :my-rating (g/game-my-rating collection-game)
     :minplayers ((g/game-attribute collection-game) :minplayers)
     :maxplayers ((g/game-attribute collection-game) :maxplayers)
     :playingtime ((g/game-attribute collection-game) :playingtime)
     :votes votes}))

(defn make-db
  [collection games]
  (let [all (map #(collection-game->game games %) collection)
        all-db (reduce
                #(assoc %1 (:id %2) %2)
                {} all)]
    (spit "resources/db.clj" (with-out-str (pp/pprint all-db)))))

(defn read-db
  []
  (read-string (slurp "resources/db.clj")))

(comment
  (def collection (g/read-collection-from-file))
  (def games (g/read-games-from-file))
  (make-db collection games)
  (def db (read-db))
  (db "25613")
  (def collection (vals db))

  (take 2 collection)
  (clojure.pprint/pp)
  ;
  )

src/bbg_api/sort_filter.clj
(ns bbg-api.sort-filter
  (:require [bbg-api.db :as db]
            [clojure.string :as s]
            [clojure.pprint :as pp]))


(defn all-games
  []
  (vals (db/read-db)))

;; 
;; Sorters
;; 
(defn game-better? [g1 g2]
  (> (:rating g1) (:rating g2)))

(defn game-shorter? [g1 g2]
  (< (:playingtime g1) (:playingtime g2)))

(defn game-shorter-and-better? [g1 g2]
  (> (/ (:rating g1) (:playingtime g1))
     (/ (:rating g2) (:playingtime g2))))

(defn rating-for-number-of-players
  "Returns a rating based on the best and recommended percentages"
  [game num-players]
  (let [_ (println (:name game) num-players)
        percentages
        (map
         (fn [pl]
           (double (/ (+  (* 2 (pl :best-votes)) (pl :recommended-votes))
                      (+ 0.0001 (* 2 (pl :best-votes)) (pl :recommended-votes) (pl :not-recommended-votes)))))
         (:votes game))
        _ (println percentages)]
    ;; Assuming all games have rating for 1 player
    ;; 1 2 3 4 4+ (5 perc)
    (if (> (count percentages) num-players)
      (nth percentages (dec num-players))
      (last percentages))))

;; 
;; Filters
;; 
(defn has-name [name]
  (fn [game]
    (s/includes? (:name game) name)))

(defn is-best-with-num-of-players
  [num-players]
  (fn [game]
    (let [best-string ((apply max-key :best-perc (:votes game)) :players)]
      (if (s/includes? best-string "+")
        ; remove '+' from the '4+'. 4+ means more than 4.
        (< num-players (read-string (s/join "" (drop-last best-string))))
        (= num-players (read-string best-string))))))

(defn is-playable-with-num-of-players
  [num threshold]
  (fn [game]
    (> (rating-for-number-of-players game num) threshold)))

(defn playingtime-between? [min max]
  (fn [game]
    (let [time (:playingtime game)]
      (and
       (number? time)
       (>= time min)
       (<= time max)))))

(defn with-number-of-players? [num]
  (fn [game] (and
              (>= (:maxplayers game) num)
              (<= (:minplayers game) num))))

(defn rating-higher-than? [rating]
  (fn [game]
    (let [game-rating (:rating game)]
      (and (not (nil? game-rating)) (>= game-rating rating)))))

;; (defn and-filters
;;   "Composes filters into one."
;;   [filter1 filter2]
;;   (fn [game] (and
;;               (filter1 game)
;;               (filter2 game))))

(defn and-filters
  "Composes filters into one."
  [& filters]
  (fn [game] (reduce
              #(and %1 %2)
              ((apply juxt filters) game))))

(comment

  (db/make-db)
  (def db-mem (all-games))
  (def game (last db-mem))
  game
  (pp/pp)

  (map
   (juxt :name :rating)
   (sort game-better?
         (filter
          (and-filters
           (with-number-of-players? 3)
           (is-best-with-num-of-players 3)
           (playingtime-between? 0 120)
           (rating-higher-than? 6))
          db-mem)))

  (map (juxt :name :id)
       (sort game-better?
             (filter
              (and-filters
               (with-number-of-players? 5)
               (is-playable-with-num-of-players 5 0.50)
               (playingtime-between? 0 240))
              db-mem)))

  (rating-for-number-of-players game 6)
  (:votes game)
  (:name game)
  (pp/pp)
  ;
  )

src/bbg_api/core.clj
(ns bbg-api.core
  (:gen-class)
  (:require [clojure.xml :as xml]
            [clojure.string :as s]))


;; 
;; Fields
;; 
(defn game-my-rating [collection-game]
  (let [rating (-> collection-game
                   :content
                   (nth 4)
                   :content
                   first
                   :attrs
                   :value
                   read-string)]
    (if (number? rating) rating nil)))

(defn game-rating [collection-game]
  (let [rating (-> collection-game
                   :content
                   (nth 4)
                   :content
                   first
                   :content
                   (nth 2)
                   :attrs
                   :value
                   read-string)]
    (if (number? rating) rating nil)))

(defn game-name [collection-game]
  (-> collection-game
      :content
      first
      :content
      first))

(defn game-attributes [collection-game]
  (-> collection-game
      :content
      (nth 4)
      :attrs))

(defn game-attribute [collection-game]
  (fn [key]
    (let [attr (game-attributes collection-game)
          value (attr key)]
      (if value (read-string value) nil))))

(defn game-playingtime [collection-game]
  ((game-attribute collection-game) :playingtime))

(defn game-maxplayers [collection-game]
  ((game-attribute collection-game) :maxplayers))

(defn game-minplayers [collection-game]
  ((game-attribute collection-game) :minplayers))

;; 
;; Collection API 
;; 
(defn fetch-collection-and-write-to-file [user-name]
  (spit "resources/collection.clj"
        (xml/parse (str "https://boardgamegeek.com/xmlapi/collection/" user-name))))

(defn read-collection-from-file []
  (:content (read-string (slurp "resources/collection.clj"))))


(defn game-id [game]
  (get-in game [:attrs :objectid]))

;; 
;; Game api
;; 
(defn api-read-game [game-id]
  (Thread/sleep 1000)
  (-> (xml/parse (str "https://boardgamegeek.com/xmlapi/boardgame/" game-id))
      :content
      first))

(defn get-games-and-write-to-file [collection]
  (spit "resources/games.clj"
        (reduce
         #(assoc %1 (get-in %2 [:attrs :objectid]) %2)
         {}
         (map api-read-game (map game-id collection)))))

(defn read-games-from-file []
  (read-string (slurp "resources/games.clj")))


"Num of players"
(defn polls-with-num-of-players-for-game [game]
  (let [tag-list (game :content)
        tag-poll (filter (fn [x] (= (x :tag) :poll)) tag-list)
        recommended (filter (fn [x] (= (get-in x [:attrs :name]) "suggested_numplayers")) tag-poll)]
    ((first recommended) :content)))

(defn votes-best-rating-per-players [data]
  (let [players (get-in data [:attrs :numplayers])
        total-votes (apply + (map (fn [x] (read-string (get-in x [:attrs :numvotes]))) (data :content)))
        best-votes (read-string (get-in (first (data :content)) [:attrs :numvotes]))
        best-perc (if (= 0 total-votes) 0 (/ best-votes total-votes))

        recommended-votes (read-string (get-in (second (data :content)) [:attrs :numvotes]))
        recommended-perc (if (= 0 total-votes) 0 (/ recommended-votes total-votes))
        not-recommended-votes (read-string (get-in (last (data :content)) [:attrs :numvotes]))
        not-recommended-perc (if (= 0 total-votes) 0 (/ not-recommended-votes total-votes))]
    {:players players
     :best-votes best-votes :best-perc best-perc
     :recommended-votes recommended-votes :recommended-perc recommended-perc
     :not-recommended-votes not-recommended-votes :not-recommended-perc not-recommended-perc}))

(defn best-with-num-of-players [games collection-game]
  (let [game (games (game-id collection-game))
        votes (map votes-best-rating-per-players (polls-with-num-of-players-for-game game))]
    ((apply max-key :best-perc votes) :players)))

(defn recommended-best-perc-players 
  "Uses a formula to find a combination of best and recommended"
  [games collection-game num-players]
  (let [game (games (game-id collection-game))
        _ (println "recommended" (game-name collection-game))
        votes (map votes-best-rating-per-players (polls-with-num-of-players-for-game game))
        percentages (map (fn [r] (double (+ (r :recommended-perc) (r :best-perc)))) votes)]
    ;; Assuming all games have rating for 1 player
    ;; 1 2 3 4 4+ (5 perc)
    (if (> (count percentages) num-players)
      (nth percentages (dec num-players))
      (last percentages))))

;; 
;; Sorters
;; 
(defn game-better? [g1 g2]
  (> (game-rating g1) (game-rating g2)))

(defn game-shorter? [g1 g2]
  (< (game-playingtime g1) (game-playingtime g2)))

(defn game-shorter-and-better? [g1 g2]
  (> (/ (game-rating g1) (game-playingtime g1))
     (/ (game-rating g2) (game-playingtime g2))))

;; 
;; Filters
;; 
(defn has-name [name]
  (fn [game]
    (s/includes? (game-name game) name)))

(defn is-best-with-num-of-players [games num]
  (fn [game]
    (let [best-string (best-with-num-of-players games game)]
      (if (s/includes? best-string "+")
        ; remove '+' from the '4+'. 4+ means more than 4.
        (< num (read-string (s/join "" (drop-last best-string))))
        (= num (read-string best-string))))))

(defn is-playable-with-num-of-players [games num threshold]
  (fn [game]
    (> (recommended-best-perc-players games game num) threshold)))

(defn playingtime-between? [min max]
  (fn [game]
    (let [time (game-playingtime game)]
      (and
       (number? time)
       (>= time min)
       (<= time max)))))

(defn with-number-of-players? [num]
  (fn [game] (and
              (>= (game-maxplayers game) num)
              (<= (game-minplayers game) num))))

(defn rating-higher-than? [rating]
  (fn [game]
    (let [game-rating (game-rating game)]
      (and (not (nil? game-rating)) (>= game-rating rating)))))

;; (defn and-filters
;;   "Composes filters into one."
;;   [filter1 filter2]
;;   (fn [game] (and
;;               (filter1 game)
;;               (filter2 game))))

(defn and-filters
  "Composes filters into one."
  [& filters]
  (fn [game] (reduce
              #(and %1 %2)
              ((apply juxt filters) game))))

(defn -main
  "I don't do a whole lot ... yet."
  [& args]
  (println "Hello, World!"))

(comment

;; collection

  ; fetch my collection from the api and write it to file in resources
  (fetch-collection-and-write-to-file "ddmits")

  ; read my collection from the file
  (def collection (read-collection-from-file))
  (def game (first collection))
  (def game (nth collection 1))
  game
  (game-name game)
  (game-id game)
  (game-rating game)

  ; read a game from api
  (def pl (api-read-game "161936"))
  (def tta (api-read-game "182028"))
  (def mar (api-read-game "276025"))
  (game-id mar)

  ; read games from file
  (def games (read-games-from-file))
  games
  (count games)
  (keys games)

  (def mar (first (filter (has-name "Maracaibo") collection)))
  mar
  (game-name mar)
  (game-id mar)
  (games (game-id mar))
  (best-with-num-of-players games mar)
  ((is-best-with-num-of-players games 3) mar)

  (def game (games (game-id mar)))
  (polls-with-num-of-players-for-game game)
  (map votes-best-rating-per-players (polls-with-num-of-players-for-game game))
  (def votes (map votes-best-rating-per-players (polls-with-num-of-players-for-game game)))
  votes
  ((apply max-key :best-perc votes) :players)

  (map (fn [r] (double (+ (r :recommended-perc) (r :best-perc)))) votes)


  (map
   (juxt game-name game-rating)
   (sort game-better?
         (filter
          (and-filters
           (with-number-of-players? 3)
           (is-best-with-num-of-players games 3)
           (playingtime-between? 0 120)
           (rating-higher-than? 6))
          collection)))

  (def gb (games "210108"))
  (polls-with-num-of-players-for-game gb)
  (def votes (map votes-best-rating-per-players (polls-with-num-of-players-for-game gb)))
  (apply max-key :best-perc votes)

  (def gbc (first (filter (has-name "Grand Bazaar") collection)))
  (best-with-num-of-players games gbc)
  ((is-best-with-num-of-players games 3) gbc)

  (def game (games (game-id mar)))
  (polls-with-num-of-players-for-game game)
  (def votes
    (map votes-best-rating-per-players (polls-with-num-of-players-for-game game)))

  (map (fn [r] (+ (r :recommended-perc) (r :best-perc))) votes)
  (nth (map (fn [r] (+ (r :recommended-perc) (r :best-perc))) votes) 3)

  ((nth (map
         (fn [r] (+ (r :recommended-perc) (r :best-perc))) votes)
        (dec 3)))


  (recommended-best-perc-players games mar 3)
  (def game10 (collection 10))
  (game-name game10)
  (recommended-best-perc-players games game10 4)
  ((is-playable-with-num-of-players games 4 0.90) game10)

  (def subcollection (take 1 collection))

  ; throws exception
  (map (juxt game-name game-id)
       (sort game-better?
             (filter
              (and-filters
               (with-number-of-players? 5)
               (is-playable-with-num-of-players games 5 0.50)
               (playingtime-between? 0 240))
              collection)))

  (def game10 (collection 10))
  (def tfm (games "167791"))
  (recommended-best-perc-players games tfm 5)
  ((is-playable-with-num-of-players games 4 0.90) game10)


  (clojure.pprint/pp)

  "names of top 10 games in collection"
  (map game-name
       (take 10
             (sort game-better? collection)))

  (take 10
        (map game-name
             (sort game-better?
                   (filter (with-number-of-players? 6) collection))))


  ;(apply  api-read-game (map game-id subcollection))

  (take 10
        (map game-name
             (sort game-better?
                   (filter (is-best-with-num-of-players games 2) collection))))

  (map game-name
       (sort game-shorter-and-better?
             (filter
              (and-filters
               (with-number-of-players? 4)
               (playingtime-between? 0 240)
               (rating-higher-than? 4))
              collection)))

  (map (juxt game-name game-id game-rating)
       (sort game-shorter-and-better?
             (filter
              (and-filters
               (with-number-of-players? 2)
               (playingtime-between? 30 240)
               (rating-higher-than? 4))
              collection)))

  (game-id (collection 0))
  (game-id (games 0))
  ;;
  )
